game-core/src/main/java/games/strategy/triplea/ResourceLoader.java

```diff
@@ -50,12 +50,15 @@ public static ResourceLoader getGameEngineAssetLoader() {
    * Returns a resource loader that will find assets in a map directory.
    */
   public static ResourceLoader getMapResourceLoader(final String mapName) {
+    @Nullable
     File atFolder = ClientFileSystemHelper.getRootFolder();
     File resourceFolder = new File(atFolder, RESOURCE_FOLDER);
 
-    while (!resourceFolder.exists() && !resourceFolder.isDirectory()) {
+    while ((atFolder != null) && !resourceFolder.exists() && !resourceFolder.isDirectory()) {
       atFolder = atFolder.getParentFile();
-      resourceFolder = new File(atFolder, RESOURCE_FOLDER);
+      if (atFolder != null) {
+        resourceFolder = new File(atFolder, RESOURCE_FOLDER);
+      }
```

Comment:
Null checking this doesn't really make sense IMO.
Because the File constructor is ok with null and the only case when this is null is when we reached the root of the file system this will effectively prevent the "last" check where RESOURCE_FOLDER might be interpreted as relative path?
This method makes it really hard to wrap your head around, and its behaviour indeed seems to be weird, as it always "finds a path", but this path doesn't necessarily exists. Luckily you added a check further down to make the behaviour less weird, but I really think this logic should be extracted into a seperate method returning an optional

Reply:
> Null checking this doesn't really make sense IMO.

Yes, you're correct.  I didn't bother to read the Javadocs. :disappointed: 

> I really think this logic should be extracted into a seperate method returning an optional

I didn't really want to touch this code because you'd be surprised how much more convoluted I discovered the resource loading system to be. :smile:  However, you make a compelling argument.

62ad009c8

I added unit tests for that new method and ended up finding a bug in the original implementation.  That last `exists()` check should have actually been an `isDirectory()` check to account for an _assets_ file at the root of the file system.

I also re-ran the tests for the IDE, `run` task, and portable build using both a headed client and a headless server, and everything seems to work as before.